File: KeySAVCoreJS/SaveBreaker.cs
using System;
using System.Linq;
using KeySAVCore.Exceptions;
using KeySAVCore.Structures;
using JS;
using DuoCode.Runtime;

namespace KeySAVCore
{
    public static class SaveBreaker
    {
        private const uint Magic = 0x42454546;
        public static readonly string[] eggnames;

        static SaveBreaker()
        {
            eggnames = new string[] {"タマゴ", "Egg", "Œuf", "Uovo", "Ei", "", "Huevo", "알"};
        }

        public static void Load(Uint8Array input, JsFunction keyGetter, JsFunction callback)
        {
            if (input.length == 0x100000)
            {
                keyGetter.invoke(BitConverter.ToUInt32(input, 0x10), BitConverter.ToUInt32(input, 0x14), (Action<Exception, SaveKey>)((Exception e, SaveKey key) => {
                    if (e == null)
                        callback.invoke(null, new SaveReaderEncrypted(input, key));
                    else
                        callback.invoke(e);
                }));
                return;
            }
            if (input.length == 0x76000 && BitConverter.ToUInt32(input, 0x75E10) == Magic)
            {
                callback.invoke(null, new SaveReaderDecrypted(input, "ORAS"));
                return;
            }
            if (input.length == 0x65600 && BitConverter.ToUInt32(input, 0x65410) == Magic)
            {
                callback.invoke(null, new SaveReaderDecrypted(input, "XY"));
                return;
            }
            if (input.length == 232 * 30 * 32)
            {
                callback.invoke(null, new SaveReaderDecrypted(input, "RAW"));
                return;
            }
            throw new NoSaveException();
        }

        // Original code by Kaphotics
        public static SaveBreakResult Break(Uint8Array break1, Uint8Array break2)
        {
            int[] offset = new int[2];
            Uint8Array empty = new Uint8Array(232);
            Uint8Array emptyekx = new Uint8Array(232);
            Uint8Array pkx = new Uint8Array(232);
            Uint8Array savkey;
            Uint8Array save1Save;
            savkey = new Uint8Array(0xB4AD4);
            string result;

            if (!Utility.SequenceEqual(break1, 0x10, break2, 0x10, 8))
            {
                return new SaveBreakResult(false, "Saves are not from the same game!\nPlease follow the instructions.", null, null);
            }

            // TODO readd upgrade logic
            if (Utility.SequenceEqual(break1, 0x80000, break2, 0x80000, 0x7F000))
            {
                save1Save = break2;
                for (int i = 0x27A00; i < 0x6CED0; ++i)
                    break2[i + 0x7F000] = (byte) (break2[i] ^ break1[i] ^ break1[i + 0x7F000]);
            }
            else if (Utility.SequenceEqual(break1, 0x1000, break2, 0x1000, 0x7F000))
            {
                save1Save = break1;
            }
            else
            {
                return new SaveBreakResult(false, "The saves are seperated by more than one save.\nPlease follow the instructions.", null, null);
            }

            Utility.Switch(ref break1, ref break2);

            #region Finding the User Specific Data: Using Valid to keep track of progress...
            // Do Break. Let's first do some sanity checking to find out the 2 offsets we're dumping from.
            // Loop through save file to find
            int fo = 0xA6A00; // Initial Offset, can tweak later.

            for (int d = 0; d < 2; d++)
            {
                // Do this twice to get both box offsets.
                for (int i = fo; i <= 0xB8F30; i += 0x10A00) 
                {
                    int err = 0;
                    // Start at findoffset and see if it matches pattern
                    if ((break1[i + 4] == break2[i + 4]) && (break1[i + 4 + 232] == break2[i + 4 + 232]))
                    {
                        // Sanity Placeholders are the same
                        for (int j = 0; j < 4; j++)
                            if (break1[i + j] == break2[i + j])
                                err++;

                        if (err < 4)
                        {
                            // Keystream ^ PID doesn't match entirely. Keep checking.
                            for (int j = 8; j < 232; j++)
                                if (break1[i + j] == break2[i + j])
                                    err++;

                            if (err < 20)
                            {
                                // Tolerable amount of difference between offsets. We have a result.
                                offset[d] = i;
                                break;
                            }
                        }
                    }
                }
                fo = offset[d] + 232 * 30;  // Fast forward out of this box to find the next.
            }

            // Now that we have our two box offsets...
            // Check to see if we actually have them.

            if ((offset[0] == 0) || (offset[1] == 0))
            {
                // We have a problem. Don't continue.
                result = "Unable to Find Box.\n";
                result += "Keystreams were NOT bruteforced!\n\nStart over and try again :(";
                return new SaveBreakResult(false, result, null, null);
            }
            else
            {
                // Let's go deeper. We have the two box offsets.
                // Chunk up the base streams.
                Uint8Array estream1 = new Uint8Array(30 * 232);
                Uint8Array estream2 = new Uint8Array(30 * 232);
                // Stuff 'em.
                for (int i = 0; i < 30; i++)    // Times we're iterating
                {
                    for (int j = 0; j < 232; j++)   // Stuff the Data
                    {
                        estream1[i * 232 + j] = break1[offset[0] + 232 * i + j];
                        estream2[i * 232 + j] = break2[offset[1] + 232 * i + j];
                    }
                }

                // Okay, now that we have the encrypted streams, formulate our EKX.
                string nick = eggnames[1];
                // Stuff in the nickname to our blank EKX.
                Uint8Array nicknamebytes = Encoding.Unicode.GetBytes(nick);
                Uint8ArrayHelper.Copy(nicknamebytes, 0, empty, 0x40, nicknamebytes.length);

                // Encrypt the Empty PKX to EKX.
                Uint8ArrayHelper.Copy(empty, 0, emptyekx, 0, 232);
                emptyekx = PKX.decrypt(emptyekx);
                // Not gonna bother with the checksum, as this empty file is temporary.

                // Sweet. Now we just have to find the E0-E3 values. Let's get our polluted streams from each.
                // Save file 1 has empty box 1. Save file 2 has empty box 2.
                Uint8Array pstream1 = new Uint8Array(30 * 232); // Polluted Keystream 1
                Uint8Array pstream2 = new Uint8Array(30 * 232); // Polluted Keystream 2
                for (int i = 0; i < 30; i++)    // Times we're iterating
                {
                    for (int j = 0; j < 232; j++)   // Stuff the Data
                    {
                        pstream1[i * 232 + j] = (byte)(estream1[i * 232 + j] ^ emptyekx[j]);
                        pstream2[i * 232 + j] = (byte)(estream2[i * 232 + j] ^ emptyekx[j]);
                    }
                }

                // Cool. So we have a fairly decent keystream to roll with. We now need to find what the E0-E3 region is.
                // 0x00000000 Encryption Constant has the D block last. 
                // We need to make sure our Supplied Encryption Constant Pokemon have the D block somewhere else (Pref in 1 or 3).

                // First, let's get out our polluted EKX's.
                byte[,] polekx = new Byte[6, 232];
                for (int i = 0; i < 6; i++)
                    for (int j = 0; j < 232; j++) // Save file 1 has them in the second box. XOR them out with the Box2 Polluted Stream
                        polekx[i, j] = (byte)(break1[offset[1] + 232 * i + j] ^ pstream2[i* 232 + j]);

                uint[] encryptionconstants = new uint[6]; // Array for all 6 Encryption Constants. 
                int valid = 0;
                for (int i = 0; i < 6; i++)
                {
                    encryptionconstants[i] = (uint)polekx[i, 0];
                    encryptionconstants[i] += (uint)polekx[i, 1] * 0x100;
                    encryptionconstants[i] += (uint)polekx[i, 2] * 0x10000;
                    encryptionconstants[i] += (uint)polekx[i, 3] * 0x1000000;
                    // EC Obtained. Check to see if Block D is not last.
                    if (PKX.getDloc(encryptionconstants[i]) != 3)
                    {
                        valid++;
                        // Find the Origin/Region data.
                        Uint8Array encryptedekx = new Uint8Array(232);
                        Uint8Array decryptedpkx = new Uint8Array(232);
                        for (int z = 0; z < 232; z++)
                            encryptedekx[z] = polekx[i, z];

                        decryptedpkx = PKX.decrypt(encryptedekx);

                        // finalize data

                        // Okay, now that we have the encrypted streams, formulate our EKX.
                        nick = eggnames[decryptedpkx[0xE3] - 1];
                        // Stuff in the nickname to our blank EKX.
                        nicknamebytes = Encoding.Unicode.GetBytes(nick);
                        Uint8ArrayHelper.Copy(nicknamebytes, 0, empty, 0x40, nicknamebytes.length);

                        // Dump it into our Blank EKX. We have won!
                        empty[0xE0] = decryptedpkx[0xE0];
                        empty[0xE1] = decryptedpkx[0xE1];
                        empty[0xE2] = decryptedpkx[0xE2];
                        empty[0xE3] = decryptedpkx[0xE3];
                        break;
                    }
                }
                #endregion

                if (valid == 0) // We didn't get any valid EC's where D was not in last. Tell the user to try again with different specimens.
                {
                    result = "The 6 supplied Pokemon are not suitable. \nRip new saves with 6 different ones that originated from your save file.\n";
                    result += "Keystreams were NOT bruteforced!\n\nStart over and try again :(";
                    return new SaveBreakResult(false, result, null, null);
                }

                else
                {
                    #region Fix up our Empty File
                    // We can continue to get our actual keystream.
                    // Let's calculate the actual checksum of our empty pkx.
                    Uint16Array empty16 = new Uint16Array(empty.buffer);
                    ushort chk = 0;
                    for (int i = 4; i < 232 / 2; i++) // Loop through the entire PKX
                        chk += empty16[i];

                    // Apply New Checksum
                    empty16[3] = chk;

                    // Okay. So we're now fixed with the proper blank PKX. Encrypt it!
                    Uint8ArrayHelper.Copy(empty, 0, emptyekx, 0, 232);
                    emptyekx = PKX.encrypt(emptyekx);

                    // Copy over 0x10-0x1F (Save Encryption Unused Data so we can track data).
                    Uint8ArrayHelper.Copy(break1, 0x10, savkey, 0, 0x8);
                    // Include empty data
                    savkey[0x10] = empty[0xE0]; savkey[0x11] = empty[0xE1]; savkey[0x12] = empty[0xE2]; savkey[0x13] = empty[0xE3];
                    // Copy over the scan offsets.
                    Uint8ArrayHelper.Copy(BitConverter.GetBytes(offset[0]), 0, savkey, 0x1C, 4);

                    for (int i = 0; i < 30; i++)    // Times we're iterating
                    {
                        for (int j = 0; j < 232; j++)   // Stuff the Data temporarily...
                        {
                            savkey[0x100 + i * 232 + j] = (byte)(estream1[i * 232 + j] ^ emptyekx[j]);
                            savkey[0x100 + (30 * 232) + i * 232 + j] = (byte)(estream2[i * 232 + j] ^ emptyekx[j]);
                        }
                    }
                    #endregion
                    // Let's extract some of the information now for when we set the Keystream filename.
                    #region Keystream Naming
                    Uint8Array data1 = new Uint8Array(232);
                    Uint8Array data2 = new Uint8Array(232);
                    for (int i = 0; i < 232; i++)
                    {
                        data1[i] = (byte)(savkey[0x100 + i] ^ break1[offset[0] + i]);
                        data2[i] = (byte)(savkey[0x100 + i] ^ break2[offset[0] + i]);
                    }
                    Uint8Array data1a = new Uint8Array(232); Uint8Array data2a = new Uint8Array(232);
                    Uint8ArrayHelper.Copy(data1, 0, data1a, 0, 232); Uint8ArrayHelper.Copy(data2, 0, data2a, 0, 232);
                    Uint8Array pkx1 = PKX.decrypt(data1);
                    Uint8Array pkx2 = PKX.decrypt(data2);
                    ushort chk1 = 0;
                    ushort chk2 = 0;
                    for (int i = 8; i < 232; i += 2)
                    {
                        chk1 += BitConverter.ToUInt16(pkx1, i);
                        chk2 += BitConverter.ToUInt16(pkx2, i);
                    }
                    if (PKX.verifyCHK(pkx1) && Convert.ToBoolean(BitConverter.ToUInt16(pkx1, 8)))
                    {
                        // Save 1 has the box1 data
                        pkx = pkx1;
                    }
                    else if (PKX.verifyCHK(pkx2) && Convert.ToBoolean(BitConverter.ToUInt16(pkx2, 8)))
                    {
                        // Save 2 has the box1 data
                        pkx = pkx2;
                    }
                    else
                    {
                        // Data isn't decrypting right...
                        for (int i = 0; i < 232; i++)
                        {
                            data1a[i] ^= empty[i];
                            data2a[i] ^= empty[i];
                        }
                        pkx1 = PKX.decrypt(data1a); pkx2 = PKX.decrypt(data2a);
                        if (PKX.verifyCHK(pkx1) && Convert.ToBoolean(BitConverter.ToUInt16(pkx1, 8)))
                        {
                            // Save 1 has the box1 data
                            pkx = pkx1;
                        }
                        else if (PKX.verifyCHK(pkx2) && Convert.ToBoolean(BitConverter.ToUInt16(pkx2, 8)))
                        {
                            // Save 2 has the box1 data
                            pkx = pkx2;
                        }
                        else
                        {
                            // Sigh...
                            return new SaveBreakResult(false, "", null, null);
                        }
                    }
                    #endregion
                }
            }
            
            if (true)
            {
                // Clear the keystream file...
                savkey.fill(0, 0x100, 0x100+232*30*31);
                savkey.fill(0, 0x40000, 0x40000+232*30*31);

                // Copy the key for the slot selector
                Uint8ArrayHelper.Copy(save1Save, 0x168, savkey, 0x80000, 4);

                // Copy the key for the other save slot
                Utility.xor(break2, offset[0], break2, offset[0]-0x7F000, savkey, 0x80004, 232*30*31);

                // Since we don't know if the user put them in in the wrong order, let's just markup our keystream with data.
                Uint8Array data1 = new Uint8Array(232);
                Uint8Array data2 = new Uint8Array(232);
                for (int i = 0; i < 31; i++)
                {
                    for (int j = 0; j < 30; j++)
                    {
                        Uint8ArrayHelper.Copy(break1, offset[0] + i * (232 * 30) + j * 232, data1, 0, 232);
                        Uint8ArrayHelper.Copy(break2, offset[0] + i * (232 * 30) + j * 232, data2, 0, 232);
                        if (data1.SequenceEqual(data2))
                        {
                            // Just copy data1 into the key file.
                            Uint8ArrayHelper.Copy(data1, 0, savkey, 0x00100 + i * (232 * 30) + j * 232, 232);
                        }
                        else
                        {
                            // Copy both datas into their keystream spots.
                            Uint8ArrayHelper.Copy(data1, 0, savkey, 0x00100 + i * (232 * 30) + j * 232, 232);
                            Uint8ArrayHelper.Copy(data2, 0, savkey, 0x40000 + i * (232 * 30) + j * 232, 232);
                        }
                    }
                }

                // Save file diff is done, now we're essentially done. Save the keystream.

                // Success
                result = "Keystreams were successfully bruteforced!\n\n";
                result += "Save your keystream now...";
                SaveKey tmp = new SaveKey(savkey);
                return new SaveBreakResult(true, result, tmp, pkx);
            }
        }
    }
}

File: save-breaker.ts
"use strict";

import SaveReader from "./save-reader";
import SaveKey from "./save-key";
import SaveReaderEncrypted from "./save-reader-encrypted";
import SaveReaderDecrypted from "./save-reader-decrypted";
import { getKeyStore } from "./key-store";
import Pkx from "./pkx";
import * as util from "./util";

const magic = 0x42454546;
export const eggnames: string[] = ["タマゴ", "Egg", "Œuf", "Uovo", "Ei", "", "Huevo", "알"];

function createNotASaveError() {
    var e = new Error("The supplied data is not a supported save type.") as any;
    e.name = "NotASaveError";
    return e;
}

export async function load(input: Uint8Array): Promise<SaveReader> {
    var view = util.createDataView(input);
    switch (input.length) {
        case 0x10009C:
        case 0x10019A:
            input = input.subarray(input.length - 0x100000);
        case 0x100000:
            var key = await getKeyStore().getSaveKey(util.getStampSav(input, 0x10));
            return new SaveReaderEncrypted(input, key);
        case 0x76000:
            if (view.getUint32(0x75E10, true) != magic)
                throw createNotASaveError();
            return new SaveReaderDecrypted(input, "ORAS");
        case 0x65600:
            if (view.getUint32(0x65410, true) != magic)
                throw createNotASaveError();
            return new SaveReaderDecrypted(input, "XY");
        case 232 * 30 * 32:
            return new SaveReaderDecrypted(input, "YABD");
        case 232 * 30 * 31:
            return new SaveReaderDecrypted(input, "PCDATA");
        case 0x70000:
            return new SaveReaderDecrypted(input, "ORASRAM");
        default:
            throw createNotASaveError();
    }
}

function upgradeKey(key: SaveKey, break1: Uint8Array, break2: Uint8Array): { result: number, pkx?: Pkx} {
    var reader1: SaveReader, reader2: SaveReader;
    var dataView1: DataView, dataView2: DataView;

    dataView1 = util.createDataView(break1);
    dataView2 = util.createDataView(break2);

    if (key.isNewKey) {
        // Scan the two saves to improve the key.
        reader1 = new SaveReaderEncrypted(break1, key); reader1.scanSlots();
        reader2 = new SaveReaderEncrypted(break2, key); reader2.scanSlots();
        // We already have a key.
        return { result: 0 };
    }

    if (util.sequenceEqual(break1, 0x80000, break2, 0x80000, 0x7f000)) {
        // We have written to slot 1 in the second save
        key.slot1Flag = dataView2.getUint32(0x168, true);
    } else if (util.sequenceEqual(break1, 0x1000, break2, 0x1000, 0x7f000)) {
        // We have written to slot 2 in the second save and as such to slot 1 in the first save
        key.slot1Flag = dataView1.getUint32(0x168, true);
    } else {
        reader1 = new SaveReaderEncrypted(break1, key); reader1.scanSlots();
        reader2 = new SaveReaderEncrypted(break2, key); reader2.scanSlots();
        // The saves are seperated by more than one save. Couldn't upgrade to a new style key.
        return {
            result: 1,
            pkx: reader1.getPkx(0) || reader2.getPkx(30)
        };
    }

    // This XORpad can encode/decode between slot 1 and slot 2 data.
    util.xor(break1, key.boxOffset, break1, key.boxOffset - 0x7f000, key.slot1Key, 0, 232*30*31);

    reader1 = new SaveReaderEncrypted(break1, key); reader1.scanSlots();
    reader2 = new SaveReaderEncrypted(break2, key); reader2.scanSlots();

    // Successfully upgraded to a new style key.
    return {
        result: 2,
        pkx: reader1.getPkx(0) || reader2.getPkx(30)
    };
}

function checkLength(file: Uint8Array) {
    let length = file.length;
    return length === 0x100000 || length === 0x10009C || length === 0x10019A;
}

export async function breakKey(break1: Uint8Array, break2: Uint8Array): Promise<string> {
    var emptyPkx = new Uint8Array(232);
    var emptyEkx = Pkx.encrypt(emptyPkx);
    var key: SaveKey;
    var boxes1: Uint8Array, boxes2: Uint8Array;
    var boxesDataView1: DataView, boxesDataView2: DataView;

    if (!checkLength(break1)) {
        let e = new Error("File 1 is not a valid save file.") as any;
        e.name = "NotASaveError";
        e.file = 1;
        throw e;
    }

    if (!checkLength(break2)) {
        let e = new Error("File 2 is not a valid save file.") as any;
        e.name = "NotASaveError";
        e.file = 2;
        throw e;
    }

    break1 = break1.subarray(break1.length % 0x100000);
    break2 = break2.subarray(break2.length % 0x100000);

    if (!util.sequenceEqual(break1, 16, break2, 16, 8)) {
        let e = new Error("The saves are not from the same game!");
        e.name = "NotSameGameError";
        throw e;
    }

    if (util.sequenceEqual(break1, break2)) {
        let e = new Error("The saves are identical.\nPlease follow the instructions.");
        e.name = "SaveIdenticalError";
        throw e;
    }

    // Let's try to upgrade an existing old style key to a new style key.
    try {
        key = await getKeyStore().getSaveKey(util.getStampSav(break1, 0x10));
        let res = upgradeKey(key, break1, break2);
        switch (res.result) {
            case 0:
                var e = new Error("You already have a key for this save.");
                e.name = "SaveKeyAlreadyExistsError";
                throw e;
            case 1:
                return "NOT_UPGRADED";
            case 2:
                return "UPGRADED";
        }
    } catch (e) {}

    key = new SaveKey(new Uint8Array(0xB4AD4));

    boxes1 = break1.subarray(0x80000, 0xFF000);
    if (util.sequenceEqual(break1, 0x80000, break2, 0x80000, 0x7F000)) {
        // We have written to only slot 1 in the second save
        boxes2 = util.xorThree(break1, 0x1000, break1, 0x80000, break2, 0x1000, 0x7F000);
    } else {
        boxes2 = break2.subarray(0x80000, 0xFF000);
    }

    boxesDataView1 = util.createDataView(boxes1);
    boxesDataView2 = util.createDataView(boxes2);

    var offset: number = undefined ;
    var potentialOffsets = [0x26A00 /* XY */, 0x37400 /* ORAS */];

    const indices = [0, 232, 464, 696, 928, 1160]; // the first six multiples of 232
    for (let i of potentialOffsets) {
        // Check that sanity placeholders are the same for all six Pokémon
        if (indices.some((j) => boxesDataView1.getUint16(i + j + 4, true) != boxesDataView2.getUint16(i + j + 4, true))) {
            continue;
        }

        // If the PID is equal for both saves this is not our offset since the Pokémon were supposed to be moved
        if (indices.some((j) => boxesDataView1.getUint32(i + j, true) == boxesDataView2.getUint32(i + j, true))) {
            continue;
        }

        let err = 0;
        for (var j = 8; j < 232; j++) {
            if (boxes1[i + j] == boxes2[i + j])
                err++;
        }

        if (err < 56) {
            offset = i + 0x80000; // Add the offset for the actual save inside the save file
            boxes1 = boxes1.subarray(i, i + 232 * 30 * 31);
            boxes2 = boxes2.subarray(i, i + 232 * 30 * 31);
            break;
        }
    }

    if (offset === undefined) {
        var e = new Error("Unable to find boxes.");
        e.name = "NoBoxesError";
        throw e;
    }

    boxesDataView1 = util.createDataView(boxes1);
    boxesDataView2 = util.createDataView(boxes2);

    // To get the keystream we need to get the complete empty ekx. It contains location data 0xE0-0xE3 and the egg name.
    // 0x00000000 Encryption Constant has the D block last.
    // We need a Pokémon with block D somewhere else so we can get the location data.
    var valid = false;
    for (let i of indices) {
        // First, let's get out our EKXs with bytes 0xE0-0xE3 random.
        var incompleteEkx = util.xorThree(boxes1, i, boxes2, i, emptyEkx, 0, 232);
        let encryptionConstant = util.createDataView(incompleteEkx).getUint32(0, true);

        // If Block D is last, the location data wouldn't be correct and we need that to fix the keystream.
        if (Pkx.getDloc(encryptionConstant) != 3) {
            valid = true;
            var incompletePkx = Pkx.decrypt(incompleteEkx);

            var nickName = eggnames[incompletePkx[0xE3] - 1];
            var nicknameBytes = util.encodeUnicode16LE(nickName);
            util.copy(nicknameBytes, 0, emptyPkx, 64, nicknameBytes.length);

            // Dump it into our Blank EKX. We have won!
            util.copy(incompletePkx, 0xE0, emptyPkx, 0xE0, 4);
            break;
        }
    }

    if (!valid) {
        // We didn't get any valid EC's where D was not in last. Tell the user to try again with different specimens.
        var e = new Error("The 6 supplied Pokemon are not suitable.");
        e.name = "PokemonNotSuitableError";
        throw e;
    }

    // This is now the complete blank pkx.
    Pkx.fixChk(emptyPkx);
    emptyEkx = Pkx.encrypt(emptyPkx);

    key.setStamp(break1);
    key.blank = emptyEkx;
    key.boxOffset = offset;

    var result = upgradeKey(key, break1, break2);
    getKeyStore().setSaveKey(key, result.pkx);
    var zeros = new Uint8Array(232);
    var ezeros = Pkx.encrypt(zeros);
    if (result.result === 2) {
        // Set the keys for slots 1-6 in boxes 1 and 2
        for (let i of indices) {
            for (let empty of [ezeros, emptyEkx]) {
                if (Pkx.verifyChk(Pkx.decrypt(util.xorThree(boxes1, i + 232 * 30, empty, 0, boxes2, i + 232 * 30, 232)))) {
                    util.copy(zeros, 0, key.boxKey1, i + 232 * 30, 232);
                    util.xor(boxes1, i + 232 * 30, empty, 0, key.boxKey2, i + 232 * 30, 232);
                    break;
                }
            }
            for (let empty of [ezeros, emptyEkx]) {
                if (Pkx.verifyChk(Pkx.decrypt(util.xorThree(boxes2, i, empty, 0, boxes1, i, 232)))) {
                    util.copy(zeros, 0, key.boxKey1, i, 232);
                    util.xor(boxes2, i, empty, 0, key.boxKey2, i, 232);
                    break;
                }
            }
        }

        return "CREATED_NEW";
    } else {
        return "CREATED_OLD";
    }
}
